barretenberg/cpp/src/barretenberg/vm2/tracegen/sha256_trace.hpp

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <list>
+
+#include "barretenberg/vm2/generated/columns.hpp"
+#include "barretenberg/vm2/simulation/events/event_emitter.hpp"
+#include "barretenberg/vm2/simulation/events/sha256_event.hpp"
+#include "barretenberg/vm2/tracegen/trace_container.hpp"
+
+namespace bb::avm2::tracegen {
+
+class Sha256TraceBuilder final {
+  public:
+    Sha256TraceBuilder(TraceContainer& trace)
+        : trace(trace)
+    {}
+    void process(const simulation::EventEmitterInterface<simulation::Sha256CompressionEvent>::Container& events);
+
+  private:
+    uint32_t row = 0;
+    TraceContainer& trace;
+
+    void into_limbs_with_witness(const uint64_t, const uint8_t b, Column c_lhs, Column c_rhs);
+    uint32_t ror_with_witness(const uint32_t val, const uint8_t shift, Column c_result, Column c_lhs, Column c_rhs);
+    uint32_t shr_with_witness(const uint32_t val, const uint8_t shift, Column c_result, Column c_lhs, Column c_rhs);
+    uint32_t compute_w_with_witness(std::array<uint32_t, 16> prev_w_helpers);
+    std::array<uint32_t, 8> compute_compression_with_witness(std::array<uint32_t, 8> state,
+                                                             uint32_t round_w,
+                                                             uint32_t round_constant,
+                                                             uint32_t row);
+    void set_helper_cols(std::array<uint32_t, 16> prev_w_helpers);
+    void set_init_state_cols(std::array<uint32_t, 8> init_state);
+    void set_state_cols(std::array<uint32_t, 8> state);
+    void compute_sha256_output(std::array<uint32_t, 8> state, std::array<uint32_t, 8> init_state);
```

Comment:
You'll want all these arrays to be `const &` unless you want to force a copy for some reason. If you do, please add a comment.